### PR TITLE
Fix browser detection

### PR DESCRIPTION
**Pull request by mmermerkaya**
Web Workers are browser too.

### CODE REVIEW

**Comment by mmermerkaya** [comment]
BTW, should we merge these `isX` functions into a single `detectEnvironment` function?

File: packages/util/src/environment.ts [hunk 1]
```diff
@@ -68,13 +67,13 @@ export function isNode(): boolean {
  * Detect Browser Environment
  */
 export function isBrowser(): boolean {
-  return typeof window !== 'undefined';
+  return typeof self === 'object' && self.self === self;
```
Comment (Feiyang1) [inline packages/util/src/environment.ts]: Just curious why checking self.self? Does some none browser env have self, but not self.self, or just to handle the case where someone defined a variable named self in none browse env? 

The intention of this function was just to check if it's a browser window env.
My intuition is that webworker probably needs a dedicated env check logic, but we don't need it anywhere. Do you see any implication of treating webworker as a browser env?

**Comment by Feiyang1** [comment]
I was thinking something along that line. We could just export the result of the function. Then you can do something like
```javascript
import { Environment, ENVIRONMENT } from `@firebase/util`
if (Environment === ENVIRONMENT.Browser) ....
```

We can also extend the function to return an object if there is other useful information. I think it's a good idea.

File: packages/util/src/environment.ts (hunk 1 above)
Comment (mmermerkaya) [reply packages/util/src/environment.ts]: To make sure that self is actually the global self, and not some random global that a developer defined.

I thought that in [the place you use this function](https://github.com/firebase/firebase-js-sdk/blob/f6afa45fe24f7adf5ea9a672100049c2ca8571db/packages/app/index.ts#L26), you'd want the result to be true for web workers as well. It's a useful check in both window and web worker contexts.

**Pull request merged**